app/models/concerns/host_searching.rb

```diff
@@ -139,6 +136,25 @@ def test_results?(_filter, _operator, value)
 
     private
 
+    def compliant_host_ids(value)
+      profiles = RequestStore.store['scoped_search_context_profiles']
+
+      TestResult.latest
+                .where(profile: profiles || Profile.all)
```

Comment:
How is this going to work without profiles context using all existing profiles, hence all hosts from all test results? The query would likely return a whole set that is in the database, wouldn't it? Considering that the ids are mapped into a string that is passed into the WHERE clause this might be a problem.

Reply:
This is a bit more delicate, we're using these IDs to reduce an already reduced (pundit-scoped) set of hosts. If you look at the other custom search methods, I'm just following a pattern. These queries are usually context free (except of the ones I introduced for some special cases) and they're only used to append an `AND WHERE` to the already running query initiated by the incoming API/GQL request.

Reply:
This picks up TestResults out of requested or all profiles. Profiles should be scope limited here, not just hosts. Other custom search methods in this concern use limited set of profiles. `Profile.all` might be the problematic part that I'm concerned about.

Here is resulting query when using REST search:
```
SELECT "test_results"."host_id"
FROM   "test_results"
       INNER JOIN "profiles"
               ON "profiles"."id" = "test_results"."profile_id"
       INNER JOIN "policies"
               ON "policies"."id" = "profiles"."policy_id"
       JOIN (SELECT "test_results"."profile_id",
                    "test_results"."host_id",
                    Max(end_time) AS end_time
             FROM   "test_results"
             GROUP  BY "test_results"."profile_id",
                       "test_results"."host_id") AS tr
         ON test_results.profile_id = tr.profile_id
            AND test_results.host_id = tr.host_id
            AND test_results.end_time = tr.end_time
WHERE  "test_results"."profile_id" IN (SELECT "profiles"."id"
                                       FROM   "profiles")
GROUP  BY "test_results"."host_id",
          policies.id,
          policies.compliance_threshold
HAVING ( Avg(test_results.score) <= 100 )
       AND ( Avg(test_results.score) < policies.compliance_threshold )
```
I'm failing to see the part that limits the scope.

Reply:
The query you described above is a helper query to return all the host IDs that are compliant. In the context of the `filter_by_compliance` method we have no contextual information (we actually dom since my hack with the `RequestStore` but that's not always there, so let's just ignore that for a moment). So the search query returns us all the host IDs that are compliant which then we use in the main query that the scoped search builds based on what this `filter_by_compliance` method returns:

```
SELECT *** FROM hosts WHERE account_id = ? AND host_id IN (scooped_search_method_kicks_in_here)
```

Reply:
> So the search query returns us all the host IDs that are compliant ...

The fact that it returns all ids and formats them as raw data to a WHERE clause is the problematic part. There can be thousands (and even more) systems. Passing it as as a subquery as mentioned [in the comment above](https://github.com/RedHatInsights/compliance-backend/pull/997#discussion_r705090927) could mitigate it. Considering that the hosts are coming from a separate database it also might become an issue when the number of compliant hosts would greatly increase.

Reply:
@vkrizan I'm definitely going to make it a subquery, I thought there's some additional issue here :see_no_evil: